commit 6e89fff225084fa685fe49438f4b7b4505dbf075
Author: Martin H <textshell-I1QKlO@neutronstar.dyndns.org>
Date:   Sun Aug 16 18:16:00 2009 +0200

    Add new options for coloring in chat dialog and system messages in muc.
    Based on patches from psi+ and Prados.
    Spooled messages and system messages share the same color, because the color difference was almost unnoticable and it's not worth adding yet another option.
    fixes #454

diff --git a/options/default.xml b/options/default.xml
index 74021f2..7b696bd 100644
--- a/options/default.xml
+++ b/options/default.xml
@@ -244,6 +244,11 @@
 							<item>Red</item>
 						</nick-colors>
 					</muc>
+					<messages comment="Message coloring.">
+						<received type="QColor" comment="Color used to indicate received messages.">#0000ff</received>
+						<sent type="QColor" comment="Color used to indicate sent messages.">#ff0000</sent>
+						<informational type="QColor" comment="Color used to indicate informational (status change, spooled) messages.">#008000</informational>
+					</messages>
 					<passive-popup>
 						<border type="QColor">#5297f9</border>
 					</passive-popup>
diff --git a/src/groupchatdlg.cpp b/src/groupchatdlg.cpp
index baed25b..8e0114d 100644
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -1458,7 +1458,7 @@ void GCMainDlg::updateLastMsgTime(QDateTime t)
 	bool doInsert = t.date() != lastMsgTime_.date();
 	lastMsgTime_ = t;
 	if (doInsert) {
-		QString color = "#00A000";
+		QString color = PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").toString();
 		ui_.log->appendText(QString("<font color=\"%1\">*** %2</font>").arg(color).arg(t.date().toString(Qt::ISODate)));
 	}
 }
@@ -1477,7 +1477,8 @@ void GCMainDlg::appendSysMsg(const QString &str, bool alert, const QDateTime &ts
 
 	updateLastMsgTime(time);
 	QString timestr = ui_.log->formatTimeStamp(time);
-	ui_.log->appendText(QString("<font color=\"#00A000\">[%1]").arg(timestr) + QString(" *** %1</font>").arg(Qt::escape(str)));
+	QString color = PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").toString();
+	ui_.log->appendText(QString("<font color=\"%1\">[%2]").arg(color, timestr) + QString(" *** %1</font>").arg(Qt::escape(str)));
 
 	if(alert)
 		doAlert();
@@ -1535,8 +1536,9 @@ void GCMainDlg::appendMessage(const Message &m, bool alert)
 		alerttagso = "<b>";
 		alerttagsc = "</b>";
 	}
-	if(m.spooled())
-		nickcolor = "#008000"; //color = "#008000";
+	if(m.spooled()) {
+		nickcolor = PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").toString();
+	}
 
 	QString timestr = ui_.log->formatTimeStamp(m.timeStamp());
 
diff --git a/src/historydlg.cpp b/src/historydlg.cpp
index 63b3b4f..eb1fb3f 100644
--- a/src/historydlg.cpp
+++ b/src/historydlg.cpp
@@ -755,9 +755,13 @@ void HistoryViewItem::paintCell(QPainter *p, const QColorGroup & cg, int column,
 {
 	QColorGroup mycg = cg;
 	if(e->originLocal())
-		mycg.setColor(QColorGroup::Text, Qt::red);
+{
+		mycg.setColor(QColorGroup::Text, PsiOptions::instance()->getOption("options.ui.look.colors.messages.sent").toString() );
+}
 	else
-		mycg.setColor(QColorGroup::Text, Qt::blue);
+{
+		mycg.setColor(QColorGroup::Text,PsiOptions::instance()->getOption("options.ui.look.colors.messages.received").toString() );
+}
 
 	if(column == 3) {
 		QBrush br;
diff --git a/src/options/opt_appearance.cpp b/src/options/opt_appearance.cpp
index b407824..b7e5c9d 100644
--- a/src/options/opt_appearance.cpp
+++ b/src/options/opt_appearance.cpp
@@ -215,6 +215,9 @@ QWidget *OptionsTabAppearanceGeneral::widget()
 	bg_color->insert(d->pb_cAnimFront);
 	bg_color->insert(d->pb_cAnimBack);
 	bg_color->insert(d->pb_cStatus);
+	bg_color->insert(d->pb_cMessageSent);
+	bg_color->insert(d->pb_cMessageReceived);
+	bg_color->insert(d->pb_cSysMsg);
 	connect(bg_color, SIGNAL(buttonClicked(QAbstractButton*)), SLOT(chooseColor(QAbstractButton*)));
 
 	QString s = tr("Specifies the text color for a contact name in the main window when that user is \"%1\".");
@@ -238,6 +241,12 @@ QWidget *OptionsTabAppearanceGeneral::widget()
 		tr("Specifies the foreground animation color for nicks."));
 	QWhatsThis::add(d->pb_cAnimBack,
 		tr("Specifies the background animation color for nicks."));
+	QWhatsThis::add(d->pb_cMessageSent,
+		tr("Specifies the color for sent messages in chat and history windows."));
+	QWhatsThis::add(d->pb_cMessageReceived,
+		tr("Specifies the color for received messages in chat and history windows."));
+	QWhatsThis::add(d->pb_cSysMsg,
+		tr("Specifies the color for informational Messages in chat windows, like status changes and offline messages."));
 
 	// Avatars
 	//QWhatsThis::add(d->ck_avatarsChatdlg,
@@ -282,6 +291,9 @@ void OptionsTabAppearanceGeneral::applyOptions()
 	PsiOptions::instance()->setOption("options.ui.look.contactlist.status-change-animation.color1", getColor(d->pb_cAnimFront));
 	PsiOptions::instance()->setOption("options.ui.look.contactlist.status-change-animation.color2", getColor(d->pb_cAnimBack));
 	PsiOptions::instance()->setOption("options.ui.look.colors.contactlist.status-messages", getColor(d->pb_cStatus));
+	PsiOptions::instance()->setOption("options.ui.look.colors.messages.received", getColor(d->pb_cMessageReceived));
+	PsiOptions::instance()->setOption("options.ui.look.colors.messages.sent", getColor(d->pb_cMessageSent));
+	PsiOptions::instance()->setOption("options.ui.look.colors.messages.informational", getColor(d->pb_cSysMsg));
 }
 
 static void restoreColor(QToolButton *button, QColor c)
@@ -315,6 +327,9 @@ void OptionsTabAppearanceGeneral::restoreOptions()
 	restoreColor(d->pb_cAnimFront, PsiOptions::instance()->getOption("options.ui.look.contactlist.status-change-animation.color1").value<QColor>());
 	restoreColor(d->pb_cAnimBack, PsiOptions::instance()->getOption("options.ui.look.contactlist.status-change-animation.color2").value<QColor>());
 	restoreColor(d->pb_cStatus, PsiOptions::instance()->getOption("options.ui.look.colors.contactlist.status-messages").value<QColor>());
+	restoreColor(d->pb_cMessageReceived, PsiOptions::instance()->getOption("options.ui.look.colors.messages.received").value<QColor>());
+	restoreColor(d->pb_cMessageSent, PsiOptions::instance()->getOption("options.ui.look.colors.messages.sent").value<QColor>());
+	restoreColor(d->pb_cSysMsg, PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").value<QColor>());
 }
 
 void OptionsTabAppearanceGeneral::setData(PsiCon *, QWidget *parentDialog)
diff --git a/src/options/opt_appearance.ui b/src/options/opt_appearance.ui
index ed86f7e..306dc58 100644
--- a/src/options/opt_appearance.ui
+++ b/src/options/opt_appearance.ui
@@ -165,6 +165,69 @@
          <string/>
         </property>
        </widget>
+       </item>
+       <item row="6" column="1" >
+       <widget class="QToolButton" name="pb_cMessageSent" >
+        <property name="sizePolicy" >
+         <sizepolicy>
+          <hsizetype>0</hsizetype>
+          <vsizetype>0</vsizetype>
+          <horstretch>0</horstretch>
+          <verstretch>0</verstretch>
+         </sizepolicy>
+        </property>
+        <property name="maximumSize" >
+         <size>
+          <width>20</width>
+          <height>20</height>
+         </size>
+        </property>
+        <property name="text" >
+         <string/>
+        </property>
+       </widget>
+      </item>
+	<item row="7" column="1">
+       <widget class="QToolButton" name="pb_cMessageReceived" >
+        <property name="sizePolicy" >
+         <sizepolicy>
+          <hsizetype>0</hsizetype>
+          <vsizetype>0</vsizetype>
+          <horstretch>0</horstretch>
+          <verstretch>0</verstretch>
+         </sizepolicy>
+        </property>
+        <property name="maximumSize" >
+         <size>
+          <width>20</width>
+          <height>20</height>
+         </size>
+        </property>
+        <property name="text" >
+         <string/>
+        </property>
+       </widget>
+      </item>
+      <item row="8" column="1">
+       <widget class="QToolButton" name="pb_cSysMsg">
+        <property name="sizePolicy" >
+         <sizepolicy>
+          <hsizetype>0</hsizetype>
+          <vsizetype>0</vsizetype>
+          <horstretch>0</horstretch>
+          <verstretch>0</verstretch>
+         </sizepolicy>
+        </property>
+        <property name="maximumSize" >
+         <size>
+          <width>20</width>
+          <height>20</height>
+         </size>
+        </property>
+        <property name="text" >
+         <string/>
+        </property>
+       </widget>
       </item>
       <item row="1" column="1" >
        <widget class="QToolButton" name="pb_cProfileBack" >
@@ -284,6 +347,27 @@
          <string>Nick animation background:</string>
         </property>
        </widget>
+	</item>
+      <item row="6" column="0">
+       <widget class="QLabel" name="tl_cMessageSent" >
+        <property name="text" >
+         <string>Sent message foreground:</string>
+        </property>
+       </widget>
+      </item>
+      <item row="7" column="0">
+       <widget class="QLabel" name="tl_cMessageReceived" >
+        <property name="text" >
+         <string>Received message foreground:</string>
+        </property>
+       </widget>
+      </item>
+      <item row="8" column="0">
+       <widget class="QLabel" name="tl_cSysMsg">
+        <property name="text">
+         <string>Informational messages in chats:</string>
+        </property>
+       </widget>
       </item>
       <item row="0" column="0" >
        <widget class="QLabel" name="tl_cProfileFore" >
@@ -498,6 +582,9 @@
   <tabstop>pb_cGroupBack</tabstop>
   <tabstop>pb_cAnimFront</tabstop>
   <tabstop>pb_cAnimBack</tabstop>
+  <tabstop>pb_cMessageSent</tabstop>
+  <tabstop>pb_cMessageReceived</tabstop>
+  <tabstop>pb_cSysMsg</tabstop>
   <tabstop>pb_cOnline</tabstop>
   <tabstop>pb_cAway</tabstop>
   <tabstop>pb_cDND</tabstop>
diff --git a/src/psichatdlg.cpp b/src/psichatdlg.cpp
index cb5dc2e..8d35935 100644
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -532,20 +532,22 @@ void PsiChatDlg::appendSysMsg(const QString &str)
 	QDateTime t = QDateTime::currentDateTime();
 	updateLastMsgTime(t);
 	QString timestr = chatView()->formatTimeStamp(t);
-	QString color = "#00A000";
+	QString color = PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").toString();
 
 	chatView()->appendText(QString("<font color=\"%1\">[%2]").arg(color, timestr) + QString(" *** %1</font>").arg(str));
 }
 
 QString PsiChatDlg::colorString(bool local, ChatDlg::SpooledType spooled) const
 {
-	if (spooled == ChatDlg::Spooled_OfflineStorage)
-		return "#008000";
+	if (spooled == ChatDlg::Spooled_OfflineStorage) {
+		return PsiOptions::instance()->getOption("options.ui.look.colors.messages.informational").toString();
+	}
 
-	if (local)
-		return "#FF0000";
+	if (local) {
+		return PsiOptions::instance()->getOption("options.ui.look.colors.messages.sent").toString();
+	}
 
-	return "#0000FF";
+	return PsiOptions::instance()->getOption("options.ui.look.colors.messages.received").toString();
 }
 
 ChatView* PsiChatDlg::chatView() const
